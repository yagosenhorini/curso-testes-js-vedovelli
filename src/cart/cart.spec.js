import Cart from "./cart";

describe("Cart", () => {
  let cart;

  let product = {
    title: "Adidas",
    price: 35388,
  };

  let product2 = {
    title: "Nike",
    price: 41872,
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe("getTotal()", () => {
    it("should return 0 when getTotal() is executed in a newly created cart", () => {
      expect(cart.getTotal().getAmount()).toEqual(0);
    });
    it("should multiply quantity and price and receive the total amount", () => {
      cart.add({
        product,
        quantity: 2,
      });
      expect(cart.getTotal().getAmount()).toEqual(70776);
    });

    it("should ensure no more than one product exists at time", () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product,
        quantity: 1,
      });
      expect(cart.getTotal().getAmount()).toEqual(35388);
    });

    it("should update total when a product gets included and then remove one", () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.remove(product);

      expect(cart.getTotal().getAmount()).toEqual(41872);
    });
  });

  describe("checkout()", () => {
    it("should return an object with the total and the list of items", () => {
      cart.add({
        product,
        quantity: 5,
      });

      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it("should return an object when summary is called", () => {
      cart.add({
        product,
        quantity: 5,
      });

      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.summary()).toMatchSnapshot();
      expect(cart.getTotal().getAmount()).toBeGreaterThan(0);
    });

    it("should return a formatted value", () => {
      cart.add({
        product,
        quantity: 5,
      });

      cart.add({
        product: product2,
        quantity: 3,
      });
      expect(cart.summary().formatted).toEqual("R$3,025.56");
    });

    it("should reset the cart when checkout() is called", () => {
      cart.add({
        product,
        quantity: 5,
      });

      cart.checkout();

      expect(cart.getTotal().getAmount()).toEqual(0);
    });
  });

  describe("special conditions", () => {
    it("should apply percentage discount above minimun", () => {
      const condition = {
        percentage: 30,
        minimun: 2,
      };

      cart.add({
        product,
        condition,
        quantity: 3,
      });

      expect(cart.getTotal().getAmount()).toEqual(74315);
    });
    it("should apply percentage discount even quantity", () => {
      const condition = {
        quantity: 2,
      };

      cart.add({
        product,
        condition,
        quantity: 4,
      });

      expect(cart.getTotal().getAmount()).toEqual(70776);
    });
    it("should apply percentage discount odd quantity", () => {
      const condition = {
        quantity: 2,
      };

      cart.add({
        product,
        condition,
        quantity: 5,
      });

      expect(cart.getTotal().getAmount()).toEqual(106164);
    });

    it("should not apply percentage discount is below or equal minimun", () => {
      const condition = {
        quantity: 2,
        percentage: 30,
      };

      cart.add({
        product,
        condition,
        quantity: 2,
      });

      expect(cart.getTotal().getAmount()).toEqual(70776);
    });

    it("should apply percentage discount odd quantity", () => {
      const condition = {
        quantity: 2,
      };

      cart.add({
        product,
        condition,
        quantity: 1,
      });

      expect(cart.getTotal().getAmount()).toEqual(35388);
    });

    it("should receive two or more conditions and determine which will be applied/First", () => {
      const condition1 = {
        percentage: 30,
        minimun: 2,
      };

      const condition2 = {
        quantity: 2,
      };

      cart.add({
        product,
        condition: [condition1, condition2],
        quantity: 5,
      });

      expect(cart.getTotal().getAmount()).toEqual(106164);
    });

    it("should receive two or more conditions and determine which will be applied/Second", () => {
      const condition1 = {
        percentage: 80,
        minimun: 2,
      };

      const condition2 = {
        quantity: 2,
      };

      cart.add({
        product,
        condition: [condition1, condition2],
        quantity: 5,
      });

      expect(cart.getTotal().getAmount()).toEqual(35388);
    });
  });
});
