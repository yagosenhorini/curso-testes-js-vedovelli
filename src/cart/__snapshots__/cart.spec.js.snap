// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object when summary is called 1`] = `
Object {
  "formatted": "R$3,025.56",
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas",
      },
      "quantity": 5,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Nike",
      },
      "quantity": 3,
    },
  ],
  "total": Object {
    "amount": 302556,
    "currency": "BRL",
    "precision": 2,
  },
}
`;

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas",
      },
      "quantity": 5,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Nike",
      },
      "quantity": 3,
    },
  ],
  "total": 302556,
}
`;
